pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -115,18 +131,67 @@ func (t *Terminator) Evict(pods []*v1.Pod) {
 			}
 		}
 	}
-	// 2. Evict in order:
-	// a. non-critical non-daemonsets
-	// b. non-critical daemonsets
-	// c. critical non-daemonsets
-	// d. critical daemonsets
-	if len(nonCriticalNonDaemon) != 0 {
-		t.evictionQueue.Add(nonCriticalNonDaemon...)
-	} else if len(nonCriticalDaemon) != 0 {
-		t.evictionQueue.Add(nonCriticalDaemon...)
-	} else if len(criticalNonDaemon) != 0 {
-		t.evictionQueue.Add(criticalNonDaemon...)
-	} else if len(criticalDaemon) != 0 {
-		t.evictionQueue.Add(criticalDaemon...)
+
+	// EvictInOrder evicts only the first list of pods which is not empty
+	// future Evict calls will catch later lists of pods that were not initially evicted
+	t.EvictInOrder(
+		nonCriticalNonDaemon,
+		nonCriticalDaemon,
+		criticalNonDaemon,
+		criticalDaemon,
+	)
+}
+
+func (t *Terminator) EvictInOrder(pods ...[]*v1.Pod) {
+	for _, podList := range pods {
+		if len(podList) > 0 {
+			// evict the first list of pods that is not empty, ignore the rest
+			t.evictionQueue.Add(podList...)
+			return
+		}
+	}
+}
+
+func (t *Terminator) DeleteExpiringPods(ctx context.Context, pods []*v1.Pod, nodeGracePeriodExpirationTime *time.Time) error {
+	for _, pod := range pods {
+		// check if the node has an expiration time and the pod needs to be deleted
+		deleteTime := t.podDeleteTimeWithGracePeriod(nodeGracePeriodExpirationTime, pod)
+		if deleteTime != nil && time.Now().After(*deleteTime) {
+			t.recorder.Publish(terminatorevents.DeletePod(pod))
+
+			// delete pod proactively to give as much of its terminationGracePeriodSeconds as possible for deletion
+			// ensure that we clamp the maximum pod terminationGracePeriodSeconds to the node's remaining expiration time in the delete command
+			opts := &client.DeleteOptions{
+				GracePeriodSeconds: ptr.Int64(int64(time.Second * time.Until(*nodeGracePeriodExpirationTime))),
```

Comment:
can we use lo instead of knative here? we've removed all instance of knative.dev/ptr and don't want to add it back in.
```suggestion
				GracePeriodSeconds: lo.ToPtr(int64(time.Second * time.Until(*nodeGracePeriodExpirationTime))),
```